src/crypto/secp156k1_provider/secp256k1_provider_impl.cpp

```diff
@@ -0,0 +1,146 @@
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    uint8_t *publicKeyPtr = publicKey.data();
+    if (i2o_ECPublicKey(key.get(), &publicKeyPtr) != publicKeyLength) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    return KeyPair{std::move(privateKey), std::move(publicKey)};
+  }
+
+  outcome::result<PublicKey> Secp256k1ProviderImpl::derivePublicKey(const PrivateKey &privateKey) const {
+    auto key = this->bytesToPrivateKey(privateKey);
+    if (key.has_error()) {
+      return key.error();
+    }
+    PublicKey publicKey{};
+    uint8_t *publicKeyPtr = publicKey.data();
+    ssize_t generatedLength = i2o_ECPublicKey(key.value().get(), &publicKeyPtr);
+    if (generatedLength != publicKey.size()) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    return publicKey;
+  }
+
+  outcome::result<Signature> Secp256k1ProviderImpl::sign(
+      gsl::span<uint8_t> message, const PrivateKey &privateKey) const {
+    common::Hash256 digest = sha256(message);
+    auto key = bytesToPrivateKey(privateKey);
+    if (key.has_error()) {
+      return key.error();
+    }
+    std::shared_ptr<ECDSA_SIG> ecSignature{
+        ECDSA_do_sign(digest.data(), digest.size(), key.value().get()),
+        ECDSA_SIG_free};
+    if (ecSignature.get() == nullptr) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    Signature signature;
+    ssize_t signatureLength = i2d_ECDSA_SIG(ecSignature.get(), nullptr);
+    if (signatureLength < 0) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    signature.resize(signatureLength);
+    uint8_t *signaturePtr = signature.data();
+    i2d_ECDSA_SIG(ecSignature.get(), &signaturePtr);
+    return signature;
+  }
+
+  outcome::result<bool> Secp256k1ProviderImpl::verify(
+      gsl::span<uint8_t> message, gsl::span<uint8_t> signature,
+      const PublicKey &publicKey) const {
+    common::Hash256 digest = sha256(message);
+    auto key = bytesToPublicKey(publicKey);
+    if (key.has_error()) {
+      return key.error();
+    }
+    int result = ECDSA_verify(0, digest.data(), digest.size(), signature.data(),
+                              signature.size(), key.value().get());
+    if (result < 0) {
+      return CryptoProviderError::SIGNATURE_VERIFICATION_FAILED;
+    }
+    return result == 1;
+  }
+
+  outcome::result<std::shared_ptr<EC_KEY>>
+  Secp256k1ProviderImpl::bytesToPrivateKey(const PrivateKey &input) const {
+    std::shared_ptr<EC_KEY> key{EC_KEY_new_by_curve_name(NID_secp256k1),
+                                EC_KEY_free};
+    if (key.get() == nullptr) {
+      return KeyGeneratorError::INTERNAL_ERROR;
+    }
+    BIGNUM *keyNum = BN_bin2bn(input.data(), input.size(), nullptr);
+    if (keyNum == nullptr) {
+      return KeyGeneratorError::INTERNAL_ERROR;
+    }
+    EC_KEY_set_private_key(key.get(), keyNum);
+    EC_GROUP *group = EC_GROUP_new_by_curve_name(NID_secp256k1);
```

Comment:
should not this pointer be freed afterwards? (triggered on that I see EC_GROUP_**new**_by_curve_name, but do not see EC_GROUP_**free**_by_curve_name if such method exists)

Reply:
Better wrap it in unique_ptr with custom deleter.

Reply:
reuse `EC_GROUP` created by `EC_KEY_new_by_curve_name`, getter is `EC_KEY_get0_group`

Reply:
Fixed to using `EC_KEY_get0_group` created by `EC_KEY_new_by_curve_name`